enhancements/project-request-limit.md

```diff
@@ -0,0 +1,345 @@
+* The source code is here: https://github.com/openshift/origin/blob/release-3.11/pkg/project/apiserver/admission/requestlimit/admission.go
+* Use  [generic-admission-server](https://github.com/openshift/generic-admission-server) to wire the validating admission webhook server. 
+* The operator is OLM enabled.
+* Ship `ProjectRequestLimit` as a RedHat operator via OperatorHub.
+* For productization, if we go through ART, integration work should target 4.5.
+* The operator should wire the OLM manifests accordingly to facilitate disconnected install (`relatedImages`).
+
+
+### Implementation Details/Notes/Constraints
+
+#### The Operand
+We build the webhook admission server to also be an extension API server, it becomes possible to aggregate it as a normal
+API server. An `APIService` object named `v1.admission.project.openshift.io` makes the API group `v1.admission.project.openshift.io/v1`
+available within and outside of the cluster via API aggregation of `kube-apiserver`. The group can be reached at 
+`/apis/admission.project.openshift.io/v1/projectrequestlimits` of the `kube-apiserver`, i.e. via the 
+`kubernetes.default.svc` service hostname inside the cluster.
+   
+The `APIService` that represents the webhook server:
+```
+apiVersion: apiregistration.k8s.io/v1
+kind: APIService
+metadata:
+  name: v1.admission.project.openshift.io
+spec:
+  group: admission.project.openshift.io
+  version: v1
+``` 
+
+The webhook server will have the following topology:
+* The webhook is deployed as a `DaemonSet` server.
+* We restrict the webhook server to the master nodes only. Apply appropriate `nodeSelector` and `tolerations` to the 
+  `PodSpec` so that the `Pods` are scheduled on to the master nodes.
+* The webhook server is reachable to the `kube-apiserver` via `kubernetes.default.svc`.
+```
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+webhooks:
+- clientConfig:
+    service:
+      # reach the webhook via the registered aggregated API
+      namespace: default
+      name: kubernetes
+      path: /apis/admission.project.openshift.io/v1/projectrequestlimits
+    caBundle: KUBE_CA
+```
+
+The call chain to the admission webhook:
+```
+`Project` request -> kube-apiserver -> kube aggregator layer inside the kube-apiserver ->
+   openshift-apiserver -> admission layer in `openshift-apiserver` -> `kube-apiserver` -> 
+     `kube-aggregator` layer inside the `kube-apiserver` -> ProjectRequestLimit webhook 
+```
+
+#### The Operator
+The operator will allow a cluster admin to:
+* Enable or disable `ProjectRequestLimit` admission webhook on a cluster.
+* Specify `limits` on `Project` create request(s) that the `ProjectRequestLimit` admission webhook can enforce.
+* Manage other lifecycle aspects of the operand.
+
+The operator will define a CRD to expose its API. The cluster admin will interact with the operator via a custom resource:
+* We treat the `ProjectRequestLimit` admission webhook as a cluster singleton.
```

Comment:
you mean the config here, don't you? Not the webhook. The webhook is a DaemonSet.

Reply:
Yes, I mean the unit of install which is the `DaemonSet`.